src/image_surface_png.rs

```diff
@@ -14,61 +16,86 @@ use ImageSurface;
 
 struct ReadEnv<'a, R: 'a + Read> {
     reader: &'a mut R,
-    error: Option<Error>,
+    io_error: Option<Error>,
+    unwind_payload: Option<Box<dyn Any + Send + 'static>>,
 }
 
 unsafe extern "C" fn read_func<R: Read>(closure: *mut c_void, data: *mut u8, len: c_uint) -> cairo_status_t {
     let read_env: &mut ReadEnv<R> = &mut *(closure as *mut ReadEnv<R>);
     let buffer = slice::from_raw_parts_mut(data, len as usize);
-    match read_env.reader.read_exact(buffer) {
-        Ok(()) => Status::Success,
-        Err(error) => {
-            read_env.error = Some(error);
+    let result = std::panic::catch_unwind(AssertUnwindSafe(|| read_env.reader.read_exact(buffer)));
+    match result {
+        Ok(Ok(())) => {
+            Status::Success
+        }
+        Ok(Err(error)) => {
+            read_env.io_error = Some(error);
+            Status::ReadError
+        }
+        Err(payload) => {
+            read_env.unwind_payload = Some(payload);
             Status::ReadError
-        },
+        }
     }.into()
 }
 
 struct WriteEnv<'a, W: 'a + Write> {
     writer: &'a mut W,
-    error: Option<Error>,
+    io_error: Option<Error>,
+    unwind_payload: Option<Box<dyn Any + Send + 'static>>,
```

Comment:
Can we have unwind and error? Maybe makes sense to use `Either` here to make it more explicit that only one of the cases can exist, but I'm fine with this here too :)

Reply:
With the current behavior it’s possible to have both. However we may want to change the behavior to not attempt another write/read if a previous one returned an error or panicked.

Reply:
Seems reasonable to have all future read/writes fail

Reply:
Done.